File: algorithmAndDataStructure/cpp/maximumSubarray/maximumSubarray.cpp
#include <iostream>
#include <vector>
#include <limits.h>

using namespace std;

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int sum = 0, maxValue = INT_MIN, numsLen = nums.size();
        for (int i = 0; i < numsLen; i++) {
            sum += nums[i];
            maxValue = max(maxValue, sum);
            if (sum < 0) {
                sum = 0;
            }
        }
        return maxValue;
    }
};

File: algorithmAndDataStructure/golang/maximumSubarray/maximumSubarray.go
package maximumSubarray

import (
	"math"
)

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func maxSubArray(nums []int) int {
	sum, maxValue := 0, math.MinInt32

	for _, num := range nums {
		sum += num
		maxValue = max(maxValue, sum)
		if sum < 0 {
			sum = 0
		}
	}
	return maxValue
}
